src/kem_api.rs

```diff
@@ -0,0 +1,150 @@
+use core::fmt::Debug;
+
+use kem::{
+    generic_array::{typenum::U32, GenericArray},
+    Decapsulator, EncappedKey, Encapsulator, SharedSecret,
+};
+use rand::{CryptoRng, RngCore};
+
+use crate::{
+    api::CryptoCiphertextBytesTypenum, crypto_kem_dec, crypto_kem_enc, crypto_kem_keypair,
+    CRYPTO_CIPHERTEXTBYTES, CRYPTO_PUBLICKEYBYTES, CRYPTO_SECRETKEYBYTES,
+};
+
+/// A struct for generating ClassicMcEliece keypairs and encapsulating a shared key.
+#[derive(Debug)]
+pub struct ClassicMcEliece;
+
+impl ClassicMcEliece {
+    pub fn keypair<R: CryptoRng + RngCore>(&self, rng: &mut R) -> (SecretKey, PublicKey) {
+        let mut sk = [0u8; CRYPTO_SECRETKEYBYTES];
+        let mut pk = [0u8; CRYPTO_PUBLICKEYBYTES];
+
+        crypto_kem_keypair(&mut pk, &mut sk, rng);
+
+        (SecretKey(sk), PublicKey(pk))
+    }
+}
+
+impl Encapsulator<Ciphertext> for ClassicMcEliece {
+    fn try_encap<R: rand::CryptoRng + rand::RngCore>(
+        &self,
+        csprng: &mut R,
+        recip_pubkey: &<Ciphertext as EncappedKey>::RecipientPublicKey,
+    ) -> Result<(Ciphertext, SharedSecret<Ciphertext>), kem::Error> {
+        let mut ciphertext = [0u8; CRYPTO_CIPHERTEXTBYTES];
+        let mut shared_secret = GenericArray::<_, _>::default();
+        crypto_kem_enc(
+            &mut ciphertext,
+            shared_secret.as_mut(),
+            &recip_pubkey.0,
+            csprng,
+        );
+        Ok((
+            Ciphertext(ciphertext),
+            SharedSecret::<Ciphertext>::new(shared_secret),
+        ))
+    }
+}
+
+/// A wrapper wrapping the bytes of a ClassicMcEliece public key.
+#[derive(Debug)]
+pub struct PublicKey([u8; CRYPTO_PUBLICKEYBYTES]);
```

Comment:
See my issue and comments in #11. I really don't think these should be treated as second class just needed to be able to implement the KEM API. I think these should replace all usage of `[u8; SOME_CONSTANT]` in the public API of this crate.

I don't think we should have a separate `ClassicMcEliece::keypair` associated function that is basically just a less flexible copy of `classic_mceliece_rust::crypto_kem_keypair`. I think we are good with only the standalone function. But it should instead deal with nice `PublicKey` and `SecretKey` types. These types can exist even with the `kem` feature turned off.

Reply:
(I think `crypto_kem_keypair` is a strange name. So I'm going to use `keypair` for the rest of this message/PR review :stuck_out_tongue: )

I would love to see something like this exposed at the crate root:
```rust
pub fn keypair<R: CryptoRng + RngCore>(
    public_key: &mut PublicKey, 
    secret_key: &mut SecretKey, 
    rng: &mut R
) -> Result<(), Error>
```

Yeah, that also involves not using `Box<dyn Error>` as return type as well. This will never reach `no_std`